in_toto/keylib.go

```diff
@@ -2,220 +2,330 @@ package in_toto
 	if err != nil {
 		return err
 	}
+	// calculate sha256 and return string representation of keyId
 	keyHashed := sha256.Sum256(keyCanonical)
-
-	// Unmarshalling the canonicalized key into the Key object would seem natural
-	// Unfortunately, our mandated canonicalization function produces a byte
-	// slice that cannot be unmarshalled by Golang's json decoder, hence we have
-	// to manually assign the values
-	k.KeyType = keyType
-	k.KeyVal = KeyVal{
-		Public: string(keyBytesStripped),
-	}
-	k.Scheme = scheme
-	k.KeyIdHashAlgorithms = keyIdHashAlgorithms
 	k.KeyId = fmt.Sprintf("%x", keyHashed)
-
 	return nil
 }
 
 /*
-VerifySignature uses the passed Key to verify the passed Signature over the
-passed data.  It returns an error if the key is not a valid RSA public key or
-if the signature is not valid for the data.
+GeneratePublicPemBlock creates a "PUBLIC KEY" PEM block from public key byte data.
+If successful it returns PEM block as []byte slice. This function should always
+succeed, if pubKeyBytes is empty the PEM block will have an empty byte block.
+Therefore only header and footer will exist.
 */
-func VerifySignature(key Key, sig Signature, data []byte) error {
-	// Create rsa.PublicKey object from DER encoded public key string as
-	// found in the public part of the keyval part of a securesystemslib key dict
-	keyReader := strings.NewReader(key.KeyVal.Public)
-	pemBytes, err := ioutil.ReadAll(keyReader)
-	if err != nil {
-		return err
+func GeneratePublicPemBlock(pubKeyBytes []byte) []byte {
+	// construct PEM block
+	publicKeyPemBlock := &pem.Block{
+		Type:    "PUBLIC KEY",
+		Headers: nil,
+		Bytes:   pubKeyBytes,
 	}
-	rsaPub, err := ParseRSAPublicKeyFromPEM(pemBytes)
-	if err != nil {
+	return pem.EncodeToMemory(publicKeyPemBlock)
+}
+
+/*
+SetKeyComponents sets all components in our key object.
+Furthermore it makes sure to remove any trailing and leading whitespaces or newlines.
+*/
+func (k *Key) SetKeyComponents(pubKeyBytes []byte, privateKeyBytes []byte, keyType string, scheme string, keyIdHashAlgorithms []string) error {
+	// assume we have a privateKey if the key size is bigger than 0
+	switch keyType {
+	case "rsa":
+		// We need to treat RSA differently, because of interoperability
+		// reasons with the securesystemslib and the in-toto python
+		// implementation
+		if len(privateKeyBytes) > 0 {
+			k.KeyVal = KeyVal{
+				Private: strings.TrimSpace(string(privateKeyBytes)),
+				Public:  strings.TrimSpace(string(GeneratePublicPemBlock(pubKeyBytes))),
+			}
+		} else {
+			k.KeyVal = KeyVal{
+				Public: strings.TrimSpace(string(pubKeyBytes)),
+			}
+		}
+	case "ed25519":
+		if len(privateKeyBytes) > 0 {
+			k.KeyVal = KeyVal{
+				Private: strings.TrimSpace(hex.EncodeToString(privateKeyBytes)),
+				Public:  strings.TrimSpace(hex.EncodeToString(pubKeyBytes)),
+			}
+		} else {
+			k.KeyVal = KeyVal{
+				Public: strings.TrimSpace(hex.EncodeToString(pubKeyBytes)),
+			}
+		}
```

Comment:
is the edgecase on the rsa case the reason for this repetition? could we probably try to increase code reuse? (e.g., by setting a function pointer to treat the private key bytes and then do the rest on a catch-all behavior for both cases.

Reply:
This is correct. The RSA key object has the whole pem block inside of the Public and Private attributes,  while the ed25519 key only has an hex encoded key string.. No idea about code re-usage. Most of this code is filling a struct. The strings.Trimspace function stays the same for all keys, but in case of RSA for example we need to generate a public pem block if we deal with a private key. That's why I meant it would make much more sense to either store everything as hex encoded string or use the pem format for everything :D But I will have another look on this, maybe I can shrink the code size a little bit more.